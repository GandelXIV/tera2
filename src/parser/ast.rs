use std::collections::HashMap;
use std::fmt;

use crate::parser::lexer::Operator;
use crate::utils::Spanned;

/// The various possible literals in variable blocks/squared brackets
#[derive(Clone, Debug, PartialEq)]
#[allow(missing_docs)]
pub enum Literal {
    Str(String),
    Integer(i64),
    Float(f64),
    Bool(bool),
}

/// An expression is the node found in variable block, kwargs and conditions.
#[derive(Clone, Debug, PartialEq)]
#[allow(missing_docs)]
pub enum Expression {
    Str(String),
    Integer(i64),
    Float(f64),
    Bool(bool),
    Ident(String),
    Array(Vec<SpannedExpression>),
    Test(Test),
    MacroCall(MacroCall),
    FunctionCall(FunctionCall),
    Expr(Operator, Vec<SpannedExpression>),
}

pub type SpannedExpression = Spanned<Expression>;

#[derive(Clone, Debug, PartialEq)]
pub struct Test {
    pub name: String,
    pub args: Vec<SpannedExpression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct MacroCall {
    pub namespace: String,
    pub name: String,
    pub kwargs: HashMap<String, SpannedExpression>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionCall {
    pub name: String,
    pub kwargs: HashMap<String, SpannedExpression>,
}

impl SpannedExpression {
    pub(crate) fn can_be_iterated_on(&self) -> bool {
        use Expression::*;
        matches!(
            self.node,
            Str(_) | Ident(_) | Array(_) | FunctionCall(_) | Expr(_, _)
        )
    }
}

impl fmt::Display for SpannedExpression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Expression::*;

        match &self.node {
            Str(i) => write!(f, "'{}'", i),
            Integer(i) => write!(f, "{}", i),
            Float(i) => write!(f, "{}", i),
            Ident(i) => write!(f, "{}", i),
            Bool(i) => write!(f, "{}", i),
            Array(vals) => {
                write!(f, "[")?;
                for (i, s) in vals.iter().enumerate() {
                    if i == vals.len() - 1 {
                        write!(f, "{}", s)?
                    } else {
                        write!(f, "{}, ", s)?
                    }
                }
                write!(f, "]")
            }
            Test(t) => {
                write!(f, "{}", t.name)?;

                if !t.args.is_empty() {
                    write!(f, "{{",)?;
                    for (i, s) in t.args.iter().enumerate() {
                        if i == t.args.len() - 1 {
                            write!(f, "{}", s)?
                        } else {
                            write!(f, "{}, ", s)?
                        }
                    }
                    write!(f, "}}",)?;
                }
                Ok(())
            }
            MacroCall(mc) => {
                write!(f, "{}::{}", mc.namespace, mc.name)?;
                write!(f, "{{",)?;
                let mut keys = mc.kwargs.keys().collect::<Vec<_>>();
                keys.sort();
                for (i, k) in keys.iter().enumerate() {
                    if i == mc.kwargs.len() - 1 {
                        write!(f, "{}={}", k, mc.kwargs[*k])?
                    } else {
                        write!(f, "{}={}, ", k, mc.kwargs[*k])?
                    }
                }
                write!(f, "}}",)?;
                Ok(())
            }
            FunctionCall(fc) => {
                write!(f, "{}", fc.name)?;
                write!(f, "{{",)?;
                let mut keys = fc.kwargs.keys().collect::<Vec<_>>();
                keys.sort();
                for (i, k) in keys.iter().enumerate() {
                    if i == fc.kwargs.len() - 1 {
                        write!(f, "{}={}", k, fc.kwargs[*k])?
                    } else {
                        write!(f, "{}={}, ", k, fc.kwargs[*k])?
                    }
                }
                write!(f, "}}",)?;
                Ok(())
            }
            Expr(op, rest) => {
                write!(f, "({}", op)?;
                for s in rest {
                    write!(f, " {}", s)?
                }
                write!(f, ")")
            }
        }
    }
}

/// Set a variable in the context `{% set val = "hey" %}`
#[derive(Clone, Debug, PartialEq)]
pub struct Set {
    /// The name for that value in the context
    pub key: String,
    /// The value to assign
    pub value: SpannedExpression,
    /// Whether we want to set the variable globally or locally
    /// global_set is only useful in loops
    pub global: bool,
}

/// A block definition
#[derive(Clone, Debug, PartialEq)]
pub struct Block {
    /// The block name
    pub name: String,
    /// The block content
    pub body: Vec<Node>,
}

/// An if/elif/else condition with their respective body
#[derive(Clone, Debug, PartialEq)]
pub struct If {
    /// First item if the if, all the ones after are elif
    pub conditions: Vec<(SpannedExpression, Vec<Node>)>,
    /// The optional `else` block
    pub otherwise: Vec<Node>,
}

/// A filter section node `{{ filter name(param="value") }} content {{ endfilter }}`
#[derive(Clone, Debug, PartialEq)]
pub struct FilterSection {
    pub name: String,
    pub kwargs: HashMap<String, SpannedExpression>,
    /// The filter body
    pub body: Vec<Node>,
}

/// A Macro definition `{% macro hello() %}...{% endmacro %}`
#[derive(Clone, Debug, PartialEq)]
pub struct MacroDefinition {
    pub name: String,
    /// The args for that macro: name -> optional default value
    pub kwargs: HashMap<String, Option<SpannedExpression>>,
    pub body: Vec<Node>,
}

/// A forloop: can be over values or key/values
#[derive(Clone, Debug, PartialEq)]
pub struct ForLoop {
    /// Name of the key in the loop (only when iterating on map-like objects)
    pub key: Option<String>,
    /// Name of the local variable for the value in the loop
    pub value: String,
    /// Expression being iterated on
    pub container: SpannedExpression,
    /// What's in the forloop itself
    pub body: Vec<Node>,
    /// The body to execute in case of an empty object in the `{% for .. %}{% else %}{% endfor %}` construct
    pub otherwise: Vec<Node>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Include {
    pub files: Vec<String>,
    pub ignore_missing: bool,
}

/// All Tera nodes that can be encountered
#[derive(Clone, Debug, PartialEq)]
pub enum Node {
    Text(String),
    VariableBlock(SpannedExpression),
    Set(Set),
    Raw(String),
    Include(Include),
    Block(Block),
    Super,
    If(If),
    ForLoop(ForLoop),
    Continue,
    Break,
    FilterSection(FilterSection),
}
