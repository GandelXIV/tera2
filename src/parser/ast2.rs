use std::collections::HashMap;
use std::fmt;

use crate::utils::Spanned;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum UnaryOperator {
    Not,
    Minus,
}

impl fmt::Display for UnaryOperator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use UnaryOperator::*;

        let val = match self {
            Minus => "-",
            Not => "not",
        };
        write!(f, "{}", val)
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum BinaryOperator {
    // math
    Mul,
    Div,
    Mod,
    Plus,
    Minus,
    FloorDiv,
    Power,

    // comparison
    LessThan,
    GreaterThan,
    LessThanOrEqual,
    GreaterThanOrEqual,
    Equal,
    NotEqual,

    // rest
    And,
    Or,
    StrConcat,
    In,
    Is,
    Pipe,
}

impl fmt::Display for BinaryOperator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use BinaryOperator::*;

        let val = match self {
            Mul => "*",
            Power => "**",
            Div => "/",
            FloorDiv => "//",
            Mod => "%",
            Plus => "+",
            Minus => "-",
            LessThan => "<",
            GreaterThan => ">",
            LessThanOrEqual => "<=",
            GreaterThanOrEqual => ">=",
            Equal => "==",
            NotEqual => "!=",
            And => "and",
            Or => "or",
            In => "in",
            Is => "is",
            StrConcat => "~",
            Pipe => "|",
        };
        write!(f, "{}", val)
    }
}

// TODO: handle my_data.blob[val].hey at compile time -> Eg no ident

/// An expression is the node found in variable block, kwargs and conditions.
#[derive(Clone, PartialEq)]
#[allow(missing_docs)]
pub enum Expression {
    Str(Spanned<String>),
    Integer(Spanned<i64>),
    Float(Spanned<f64>),
    Bool(Spanned<bool>),
    Ident(Spanned<String>),
    Array(Spanned<Array>),
    Test(Spanned<Test>),
    MacroCall(Spanned<MacroCall>),
    FunctionCall(Spanned<FunctionCall>),
    UnaryOperation(Spanned<UnaryOperation>),
    BinaryOperation(Spanned<BinaryOperation>),
}

#[derive(Clone, PartialEq)]
pub struct UnaryOperation {
    pub op: UnaryOperator,
    pub expr: Expression,
}

impl fmt::Debug for UnaryOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({} {:?})", self.op, self.expr)
    }
}

#[derive(Clone, PartialEq)]
pub struct BinaryOperation {
    pub op: BinaryOperator,
    pub left: Expression,
    pub right: Expression,
}

impl fmt::Debug for BinaryOperation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({} {:?} {:?})", self.op, self.left, self.right)
    }
}

#[derive(Clone, PartialEq)]
pub struct Array {
    pub items: Vec<Expression>,
}

impl fmt::Debug for Array {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "[")?;
        for (i, s) in self.items.iter().enumerate() {
            if i == self.items.len() - 1 {
                write!(f, "{:?}", s)?
            } else {
                write!(f, "{:?}, ", s)?
            }
        }
        write!(f, "]")
    }
}

#[derive(Clone, PartialEq)]
pub struct Test {
    pub name: String,
    pub args: Vec<Expression>,
}

impl fmt::Debug for Test {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)?;

        if !self.args.is_empty() {
            write!(f, "{{",)?;
            for (i, s) in self.args.iter().enumerate() {
                if i == self.args.len() - 1 {
                    write!(f, "{:?}", s)?
                } else {
                    write!(f, "{:?}, ", s)?
                }
            }
            write!(f, "}}",)?;
        }
        Ok(())
    }
}

#[derive(Clone, PartialEq)]
pub struct MacroCall {
    pub namespace: String,
    pub name: String,
    pub kwargs: HashMap<String, Expression>,
}

impl fmt::Debug for MacroCall {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}::{}", self.namespace, self.name)?;
        write!(f, "{{",)?;
        let mut keys = self.kwargs.keys().collect::<Vec<_>>();
        keys.sort();
        for (i, k) in keys.iter().enumerate() {
            if i == self.kwargs.len() - 1 {
                write!(f, "{}={:?}", k, self.kwargs[*k])?
            } else {
                write!(f, "{}={:?}, ", k, self.kwargs[*k])?
            }
        }
        write!(f, "}}",)
    }
}

#[derive(Clone, PartialEq)]
pub struct FunctionCall {
    pub name: String,
    pub kwargs: HashMap<String, Expression>,
}

impl fmt::Debug for FunctionCall {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)?;
        write!(f, "{{",)?;
        let mut keys = self.kwargs.keys().collect::<Vec<_>>();
        keys.sort();
        for (i, k) in keys.iter().enumerate() {
            if i == self.kwargs.len() - 1 {
                write!(f, "{}={:?}", k, self.kwargs[*k])?
            } else {
                write!(f, "{}={:?}, ", k, self.kwargs[*k])?
            }
        }
        write!(f, "}}",)
    }
}

impl fmt::Debug for Expression {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Expression::*;

        match self {
            Str(i) => write!(f, "'{}'", **i),
            Integer(i) => write!(f, "{}", **i),
            Float(i) => write!(f, "{}", **i),
            Ident(i) => write!(f, "{}", **i),
            Bool(i) => write!(f, "{}", **i),
            Array(i) => write!(f, "{:?}", **i),
            Test(i) => write!(f, "{:?}", **i),
            MacroCall(i) => write!(f, "{:?}", **i),
            FunctionCall(i) => write!(f, "{:?}", **i),
            UnaryOperation(i) => write!(f, "{:?}", **i),
            BinaryOperation(i) => write!(f, "{:?}", **i),
        }
    }
}

// /// Set a variable in the context `{% set val = "hey" %}`
// #[derive(Clone, Debug, PartialEq)]
// pub struct Set {
//     /// The name for that value in the context
//     pub key: String,
//     /// The value to assign
//     pub value: SpannedExpression,
//     /// Whether we want to set the variable globally or locally
//     /// global_set is only useful in loops
//     pub global: bool,
// }
//
// /// A block definition
// #[derive(Clone, Debug, PartialEq)]
// pub struct Block {
//     /// The block name
//     pub name: String,
//     /// The block content
//     pub body: Vec<Node>,
// }
//
// /// An if/elif/else condition with their respective body
// #[derive(Clone, Debug, PartialEq)]
// pub struct If {
//     /// First item if the if, all the ones after are elif
//     pub conditions: Vec<(SpannedExpression, Vec<Node>)>,
//     /// The optional `else` block
//     pub otherwise: Vec<Node>,
// }
//
// /// A filter section node `{{ filter name(param="value") }} content {{ endfilter }}`
// #[derive(Clone, Debug, PartialEq)]
// pub struct FilterSection {
//     pub name: String,
//     pub kwargs: HashMap<String, SpannedExpression>,
//     /// The filter body
//     pub body: Vec<Node>,
// }
//
// /// A Macro definition `{% macro hello() %}...{% endmacro %}`
// #[derive(Clone, Debug, PartialEq)]
// pub struct MacroDefinition {
//     pub name: String,
//     /// The args for that macro: name -> optional default value
//     pub kwargs: HashMap<String, Option<SpannedExpression>>,
//     pub body: Vec<Node>,
// }
//
// /// A forloop: can be over values or key/values
// #[derive(Clone, Debug, PartialEq)]
// pub struct ForLoop {
//     /// Name of the key in the loop (only when iterating on map-like objects)
//     pub key: Option<String>,
//     /// Name of the local variable for the value in the loop
//     pub value: String,
//     /// Expression being iterated on
//     pub container: SpannedExpression,
//     /// What's in the forloop itself
//     pub body: Vec<Node>,
//     /// The body to execute in case of an empty object in the `{% for .. %}{% else %}{% endfor %}` construct
//     pub otherwise: Vec<Node>,
// }
//
// #[derive(Clone, Debug, PartialEq)]
// pub struct Include {
//     pub files: Vec<String>,
//     pub ignore_missing: bool,
// }
//
// /// All Tera nodes that can be encountered
// #[derive(Clone, Debug, PartialEq)]
// pub enum Node {
//     Content(String),
//     VariableBlock(SpannedExpression),
//     Set(Set),
//     Raw(String),
//     Include(Include),
//     Block(Block),
//     Super,
//     If(If),
//     ForLoop(ForLoop),
//     // Do those need to be in the AST?
//     Continue,
//     Break,
//     FilterSection(FilterSection),
// }

#[derive(Clone, PartialEq)]
pub enum Node {
    Content(String),
    Expression(Expression),
}

impl fmt::Debug for Node {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use Node::*;

        match self {
            Node::Content(s) =>  write!(f, "{:?}", s),
            Node::Expression(e) => write!(f, "{:?}", e),
        }
    }
}