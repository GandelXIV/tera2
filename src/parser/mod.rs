mod parser;

mod ast;
mod lexer;

pub use parser::Parser;

#[cfg(test)]
mod tests;
