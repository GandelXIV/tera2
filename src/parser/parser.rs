use crate::errors::{Error, ErrorKind, TeraResult};
use crate::parser::ast2::{Array, Expression, UnaryOperation};
use crate::parser::ast2::{BinaryOperator, Node, UnaryOperator};
use crate::parser::lexer2::{tokenize, Token};
use crate::utils::{Span, Spanned};

// From https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html

fn unary_binding_power(op: UnaryOperator) -> ((), u8) {
    use UnaryOperator::*;

    match op {
        Not => ((), 3),
        Minus => ((), 7),
    }
}

fn binary_binding_power(op: BinaryOperator) -> (u8, u8) {
    use BinaryOperator::*;

    match op {
        And | Or => (1, 2),
        In | Is => (3, 4),
        Pipe => (5, 6),
        Equal | NotEqual | LessThan | LessThanOrEqual | GreaterThan | GreaterThanOrEqual => (7, 8),
        Plus | Minus => (11, 12),
        Mul | Div | Mod | StrConcat | FloorDiv | Power => (13, 14),
    }
}

macro_rules! expect_token {
    ($parser:expr, $expectation:expr) => {{
        match $parser.next_or_error()? {
            Some(rv) => Ok(rv),
            None => Err(Error::new(ErrorKind::SyntaxError, "todo simple expect")),
        }
    }};
    ($parser:expr, $match:pat, $expectation:expr) => {{
        match $parser.next_or_error()? {
            (token, span) if matches!(token, $match) => Ok((token, span)),
            _ => Err(Error::new(ErrorKind::SyntaxError, "todo match expect")),
        }
    }};
}

// TODO: double check what is actually reserved so we can't re-assign them
const RESERVED_NAMES: [&str; 6] = ["true", "True", "false", "False", "loop", "self"];

pub struct Parser<'a> {
    source: &'a str,
    lexer: Box<dyn Iterator<Item = Result<(Token<'a>, Span), Error>> + 'a>,
    current: Option<Result<(Token<'a>, Span), Error>>,
    // We keep track of the current span TODO
    current_span: Span,
    // filled when we encounter a {% extends %}, we don't need to keep the extends node in the AST
    pub parent: Option<String>,
    // (file, namespace)
    pub macro_imports: Vec<(String, String)>,
}

impl<'a> Parser<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            source,
            lexer: Box::new(tokenize(source).peekable()),
            current: None,
            current_span: Span::default(),
            parent: None,
            macro_imports: Vec::new(),
        }
    }

    fn current(&mut self) -> Result<Option<(&Token<'a>, &Span)>, Error> {
        if self.current.is_none() {
            self.next()?;
        }

        match self.current {
            Some(Ok(ref tok)) => Ok(Some((&tok.0, &tok.1))),
            Some(Err(_)) => Err(self.current.take().unwrap().unwrap_err()),
            None => Ok(None),
        }
    }

    fn next(&mut self) -> TeraResult<Option<(Token<'a>, Span)>> {
        let cur = self.current.take();
        self.current = self.lexer.next();
        if let Some(Ok((_, ref span))) = cur {
            self.current_span = span.clone();
        }

        cur.transpose()
    }

    fn next_or_error(&mut self) -> TeraResult<(Token<'a>, Span)> {
        match self.next()? {
            None => {
                todo!("handle error message")
            }
            Some(c) => Ok(c),
        }
    }

    fn expect(&mut self, expected_token: Token, message: &str) -> TeraResult<()> {
        let (current_token, _) = self.next_or_error()?;
        if current_token != expected_token {
            return Err(Error::new(ErrorKind::SyntaxError, &format!("Expected {} but got {:?}", message, current_token)));
        }

        Ok(())
    }

    fn peek(&mut self) {}

    /// An ident can have multiple forms:
    /// `hey`
    /// `hey.ho`
    /// `hey['ho']`
    /// `hey[ho]`
    /// `hey[ho].name`
    /// `hey[1]`
    fn parse_ident(&mut self) -> TeraResult<String> {

        Ok(String::new())
    }


    // TODO: make sure it handles trailing commas
    fn parse_array(&mut self) -> TeraResult<Vec<Expression>> {
        let mut vals = Vec::new();

        loop {
            if matches!(self.current()?, Some((Token::RightBracket, _))) {
                break;
            }

            if !vals.is_empty() {
                self.expect(Token::Comma, ",")?;
            }

            vals.push(self.parse_expression(0)?);
        }

        Ok(vals)
    }

    fn parse_expression(&mut self, min_bp: u8) -> TeraResult<Expression> {
        let (token, mut span) = self.next_or_error()?;

        let lhs = match token {
            Token::Integer(i) => Expression::Integer(Spanned::new(i, span)),
            Token::Float(f) => Expression::Float(Spanned::new(f, span)),
            Token::String(s) => Expression::Str(Spanned::new(s.to_owned(), span)),
            Token::Ident("true") | Token::Ident("True") => Expression::Bool(Spanned::new(true, span)),
            Token::Ident("false") | Token::Ident("False") => Expression::Bool(Spanned::new(false, span)),
            Token::Minus | Token::Ident("not") => {
                let op = match token {
                    Token::Minus => UnaryOperator::Minus,
                    Token::Ident("not") => UnaryOperator::Not,
                    _ => unreachable!(),
                };
                let (_, r_bp) = unary_binding_power(op);
                let expr = self.parse_expression(r_bp)?;
                span.expand(&self.current_span);
                Expression::UnaryOperation(Spanned::new(UnaryOperation {op, expr}, span))
            }
            Token::Ident(ident) => {

            }
            Token::LeftBracket => {
                let items = self.parse_array()?;
                span.expand(&self.current_span);
                Expression::Array(Spanned::new(Array { items }, span))
            }
            _ => todo!("finish expression: {:?}", token),
        };

        Ok(lhs)
    }

    fn parse_until<F: Fn(&Token) -> bool>(&mut self, end_check_fn: F) -> TeraResult<Vec<Node>> {
        let mut nodes = Vec::new();

        while let Some((token, ref span)) = self.next()? {
            match token {
                Token::Content(c) => {
                    // We have pushed an empty content to replace comment
                    if !c.is_empty() {
                        nodes.push(Node::Content(c.to_owned()));
                    }
                }
                Token::VariableStart(_) => {
                    let expr = self.parse_expression(0)?;
                    expect_token!(self, Token::VariableEnd(..), "%}")?;
                    nodes.push(Node::Expression(expr));
                }
                Token::TagStart(_) => {}
                _ => unreachable!("Something wrong happened while lexing/parsing"),
            }
        }

        Ok(nodes)
    }

    pub fn parse(&mut self) -> TeraResult<Vec<Node>> {
        // get the first token
        self.next()?;
        let nodes = self.parse_until(|_| false)?;
        Ok(nodes)
    }
}
