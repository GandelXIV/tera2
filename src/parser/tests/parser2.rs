use crate::parser::ast2::Node;
use crate::parser::parser::Parser;

#[test]
fn test_parser_success() {
    insta::glob!("parser_inputs/success/*.txt", |path| {
        let contents = std::fs::read_to_string(path).unwrap();
        // TODO: remove nodes that aren't expressions to remove the `\n` in between each expr
        let nodes = &Parser::new(&contents).parse().unwrap();
        let mut expr_nodes = Vec::with_capacity(nodes.len());
        for node in nodes {
            match node {
                Node::Expression(_) => {
                    expr_nodes.push(node);
                }
                _ => ()
            }
        }
        insta::assert_debug_snapshot!(&expr_nodes);
    });
}

// #[test]
// fn test_lexer_errors() {
//     insta::glob!("parser_inputs/errors/*.txt", |path| {
//         let contents = std::fs::read_to_string(path).unwrap();
//         let res: Result<Vec<_>, _> = tokenize(&contents).collect();
//         assert!(res.is_err());
//         insta::assert_debug_snapshot!(res.unwrap_err());
//     });
// }
