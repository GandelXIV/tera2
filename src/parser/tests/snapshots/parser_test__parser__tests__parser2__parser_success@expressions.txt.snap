---
source: src/parser/tests/parser2.rs
assertion_line: 8
expression: "&expr_node"
input_file: src/parser/tests/parser_inputs/success/expressions.txt

---
1
